rpc/src/evm_state.rs

```diff
@@ -75,21 +81,32 @@ impl EvmBlockState {
     }
 
     pub async fn account_state(
-        &self,
+        &mut self,
         address_hash: B256,
     ) -> Result<Option<AccountState>, EvmStateError> {
-        self.traverse_trie(
-            self.block_header.state_root,
-            address_hash,
-            |path, node_hash| {
-                StateContentKey::AccountTrieNode(AccountTrieNodeKey { path, node_hash })
-            },
-        )
-        .await
+        let account_state = self
+            .traverse_trie::<AccountState>(
+                self.block_header.state_root,
+                address_hash,
+                |path, node_hash| {
+                    StateContentKey::AccountTrieNode(AccountTrieNodeKey { path, node_hash })
+                },
+            )
+            .await?;
+
+        if let Some(account_state) = &account_state {
+            if account_state.code_hash != KECCAK_EMPTY {
+                self.code_hash_to_address_hash
+                    .entry(account_state.code_hash)
+                    .or_insert(address_hash);
+            }
+        }
```

Comment:
Shouldn't this cache code be in `async fn basic_async(&mut self, address: Address) -> Result<Option<AccountInfo>, Self::Error> {`

Multiple contracts can have the same bytecode, but since eth_call is read only? We are probably good. I am assuming eth_call can't run self-destruct or create bytecode or anything

Reply:
> Shouldn't this cache code be in `async fn basic_async(&mut self, address: Address) -> Result<Option<AccountInfo>, Self::Error> {`

Considering that `basic_async` calls this function, I would say that this is better place (since every path that reads account state has to go through here).
But I can see the argument of being there as well (since this cache is only used by `code_by_hash_async`). I would leave it here for now, but I can move it down if you have strong opinion on this.

> Multiple contracts can have the same bytecode, but since eth_call is read only? We are probably good. I am assuming eth_call can't run self-destruct or create bytecode or anything

The way I see `EvmBlockState`: It's given a block header, and it fetches the state trie from portal network at that specific block. In the context of Evm and execution, it's the same as reading the state from a db.
So the entire evm execution is read-only from the db point of view (and state is only committed later).

With that being said, `eth_call` can both call self-destruct, create contracts, etc. but all of that is handled by evm. The `EvmBlockState` doesn't have to care about any of that.

---

Regarding multiple contracts having the same bytecode: in order to fetch it from the portal network, we only need to know one account that has it (regardless of which one it is, since they are identical). The reason we have `address_hash` in the content_key is to avoid hot-spot on portal network for the bytecode that is shared across many contracts (e.g. erc20 token, uniswap dex pool).